### DISCUSSION TITLE
Conditionally show Firebase Messaging notification depending on foreground state or content?

### DISCUSSION THREAD

**Discussion by r0hin** [capawesome-team/capacitor-firebase#397] [Q&A] [answered]
Using Firebase Cloud Messaging, upon receiving notification via FCM token, the notification is shown regardless of whether or not the app is in the foreground.

Is there any existing functionality to disable notifications when the app is open and instead use an in-app notification?

Thanks!

**Comment by robingenz** [comment]
Thank you for your request. What platform do you mean? iOS? 
Basically, you can simply send a data notification (data notifications are always hidden) and then create your own implementation using the [Local Notifications](https://capacitorjs.com/docs/apis/local-notifications) plugin.

**Reply by r0hin** [reply]
Yup – iOS. Local notifications would be great for in-app notifications, but wouldn't this cause both in-app notifications and regular iOS notifications to happen at the same time? I'd like to preserve functionality for notifications when app is not in foreground.

**Reply by robingenz** [reply, marked as answer]
> but wouldn't this cause both in-app notifications and regular iOS notifications to happen at the same time

As said, data notifications are not displayed to the user. They wake up the app for a short moment and you have time to create a local notification instead according to your own rules.

**Reply by r0hin** [reply]
I understand now! Thanks for your insight, will start building now :)!

**Reply by robingenz** [reply]
Probably you could also solve this within the plugin as features. Feel free to create a feature request and I'll take a look at it as soon as I have time. But in the meantime you can test the suggested solution.

**Reply by r0hin** [reply]
I've been testing the solution and in-app notifs work great. However, when the app is closed, push notifications don't seem to show. Were you thinking of an implementation like this?:

```
FirebaseMessaging.addListener(`notificationReceived`, async (notification) => {
    const state = await App.getState();
    if (!state.isActive) { // App is closed
      await LocalNotifications.schedule({
        notifications: [
          {
            title: notification.notification.data.title,
            body: notification.notification.data.body,
            id: new Date().getTime(),
          }
        ],
      })
    }
  else {
    ... in-app
  }
})
```

The FCM message object looks like this:
```
{
    "token": receiverToken,
    "data": {
      "title": `${senderName}`,
      "body": `${text}`,
      "sender": `${sender}`
    },
    "apns": {
      "payload": {
        "aps": {
          "contentAvailable": 1,
          "badge": 0,
        }
      }
    }
  }
```

The LocalNotifications plugin also works great – it seems like a matter of getting the notificationReceived listener to fire (possibly related to https://github.com/capawesome-team/capacitor-firebase/issues/178 or https://github.com/capawesome-team/capacitor-firebase/issues/244 ?)

**Reply by robingenz** [reply]
> However, when the app is closed, push notifications don't seem to show.

Did you enable the `Background Mode` capability and select the `Remote Notifications` checkbox (see [here](https://capawesome.io/blog/the-push-notifications-guide-for-capacitor/#enable-push-notification-capabilities))?

And what do you mean by `closed`? App in the background or killed?

**Comment by r0hin** [comment]
> Did you enable the `Background Mode` capability and select the `Remote Notifications` checkbox (see [here](https://capawesome.io/blog/the-push-notifications-guide-for-capacitor/#enable-push-notification-capabilities))?

Yes, both capabilities are enabled

>And what do you mean by closed? App in the background or killed?

In all non-foreground states (both background + killed), I can't get system notifications to go through.

**Reply by robingenz** [reply]
It should work in the background. I'll try to implement #398 in a timely manner.

**Reply by r0hin** [reply]
Incredible — thank you! I’ll continue investigating in the meantime.

**Reply by r0hin** [reply]
Not quite sure if this is what fixed it, but:
```
let state = UIApplication.shared.applicationState
if (state == .background || state == .inactive) {
    NotificationCenter.default.post(name: Notification.Name.init("didReceiveRemoteNotification"), object: completionHandler, userInfo: userInfo)
}
```

within
```
 func application(_ application: UIApplication, didReceiveRemoteNotification userInfo: [AnyHashable : Any], fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) {
```

seemed to work :)

**Reply by robingenz** [reply]
That is a very elegant solution. I should have thought of that. Thanks for sharing!
